src/main/java/org/zalando/nakadi/util/TimeLogger.java

```diff
@@ -0,0 +1,130 @@
+package org.zalando.nakadi.util;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+
+public class TimeLogger {
+    private static class LogCollector {
+        private long[] stepStarts;
+        private String[] stepNames;
+        private int currentStep;
+
+        LogCollector() {
+            this.stepStarts = new long[10];
+            this.stepNames = new String[this.stepStarts.length];
+            this.currentStep = 0;
+        }
+
+        private void reset() {
+            this.currentStep = 0;
+        }
+
+        void addRecord(final String message) {
+            if (currentStep >= this.stepStarts.length) {
+                this.stepStarts = Arrays.copyOf(this.stepStarts, this.stepStarts.length + 10);
+                this.stepNames = Arrays.copyOf(this.stepNames, this.stepStarts.length);
+            }
+            this.stepStarts[currentStep] = System.currentTimeMillis();
+            this.stepNames[currentStep] = message;
+            currentStep++;
+        }
+
+        String describe() {
+            final StringBuilder builder = new StringBuilder("[" + stepNames[0] + "]");
+            for (int i = 1; i < currentStep - 1; ++i) {
+                builder.append(" [")
+                        .append(stepNames[i])
+                        .append("=")
+                        .append(stepStarts[i + 1] - stepStarts[i])
+                        .append("]");
+            }
+            return builder.toString();
+        }
+    }
+
+    private static final ThreadLocal<LogCollector[]> THREAD_COLLECTORS = new ThreadLocal<>();
+    private static final LinkedList<LogCollector> COLLECTOR_POOL = new LinkedList<>();
+
+    public static void startMeasure(final String heading, final String stepName) {
+        LogCollector collector;
+        synchronized (COLLECTOR_POOL) {
+            collector = COLLECTOR_POOL.poll();
+        }
+        if (null == collector) {
```

Comment:
Looks like it is better to initialise it under the lock, otherwise you create several LogConnector instances.